subsys/pmci/mctp/mctp_i2c_gpio_target.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2025 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include "zephyr/pmci/mctp/mctp_i2c_gpio_common.h"
+#include <zephyr/sys/__assert.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/pmci/mctp/mctp_i2c_gpio_target.h>
+#include <crc-16-ccitt.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mctp_i2c_gpio_target, CONFIG_MCTP_LOG_LEVEL);
+
+int mctp_i2c_gpio_target_write_requested(struct i2c_target_config *config)
+{
+
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+
+	if (b->rxtx || b->reg_addr == MCTP_I2C_GPIO_INVALID_ADDR) {
+		/* Reset our state */
+		b->reg_addr = MCTP_I2C_GPIO_INVALID_ADDR;
+		b->rxtx = false;
+		b->rx_idx = 0;
+	}
+
+	return 0;
+}
+
+int mctp_i2c_gpio_target_write_received(struct i2c_target_config *config, uint8_t val)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+	int ret = 0;
+
+	switch (b->reg_addr) {
+	case MCTP_I2C_GPIO_INVALID_ADDR:
+		b->rxtx = false;
+		b->reg_addr = val;
+		break;
+	case MCTP_I2C_GPIO_RX_MSG_LEN_ADDR:
+		b->rxtx = true;
+		b->rx_pkt = mctp_pktbuf_alloc(&b->binding, (size_t)val);
+		break;
+	case MCTP_I2C_GPIO_RX_MSG_ADDR:
+		b->rxtx = true;
+		b->rx_pkt->data[b->rx_idx] = val;
+		b->rx_idx += 1;
+
+		/* buffer full */
+		if (b->rx_idx >= b->rx_pkt->size) {
+			ret = -ENOMEM;
+		}
+		break;
+	default:
+		LOG_ERR("Write when reg_addr is %d", b->reg_addr);
+		ret = -EIO;
+		break;
+	}
+
+	return ret;
+}
+
+int mctp_i2c_gpio_target_read_requested(struct i2c_target_config *config, uint8_t *val)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+	uint8_t pkt_len;
+	int ret = 0;
+
+	switch (b->reg_addr) {
+	case MCTP_I2C_GPIO_TX_MSG_LEN_ADDR:
+		b->rxtx = true;
+		if (b->tx_pkt == NULL) {
+			LOG_WRN("empty packet?");
+			pkt_len = 0;
+		} else {
+			pkt_len = (uint8_t)(b->tx_pkt->end - b->tx_pkt->start);
+		}
+		*val = pkt_len;
+		break;
+	case MCTP_I2C_GPIO_TX_MSG_ADDR:
+		b->rxtx = true;
+		*val = b->tx_pkt->data[b->tx_pkt->start];
+		b->tx_idx = b->tx_pkt->start;
+		break;
+	default:
+		LOG_WRN("invalid rre reg %d", b->reg_addr);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+int mctp_i2c_gpio_target_read_processed(struct i2c_target_config *config, uint8_t *val)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+
+	b->tx_idx += 1;
+
+	if (b->reg_addr != MCTP_I2C_GPIO_TX_MSG_ADDR) {
+		goto out;
+	}
+
+	if (b->tx_idx > b->tx_pkt->end) {
```

Comment:
I think?

```suggestion
	if (b->tx_idx >= b->tx_pkt->end) {
```

Reply:
Why do you think this?

Reply:
I was under the impression pkt_end was the first byte _past_ the end of valid data?